src/rospkg/rospack.py

```diff
@@ -391,6 +392,36 @@ def stack_of(self, package):
             else:
                 d = os.path.dirname(d)
 
+    def get_licenses(self, pkg_name, implicit=True):
+        """
+        @summary: Return a list of licenses and the packages in the dependency tree
+            for the given package. Special value 'ERR' is used as the license for the
+            packages that license was not detected for.
+        @param pkg_name: Name of the package the dependency tree begins from.
+        @return OrderedDict of license name and a list of packages.
+        @rtype { k, [d] }
+        @raise ResourceNotFound
+        """
+        license_dict = defaultdict(list)
+
+        self.get_depends(name=pkg_name, implicit=implicit)
+
+        for p_name, manifest in self._manifests.items():
+            for license in manifest.licenses:
+                license_dict[license].append(p_name)
+
+        # Traverse for Non-ROS, system packages
+        pkgnames_rosdep = self.get_rosdeps(package=pkg_name, implicit=implicit)
+        for pkgname_rosdep in pkgnames_rosdep:
+            license_dict[None].append(pkgname_rosdep)
```

Comment:
From https://github.com/ros-infrastructure/rospkg/pull/133#discussion_r270968864
> Why should this function combine both - the results from `get_depends` as well as `get_rosdeps` - into a single result? That seems very use case specific.
> 
> In general since the function only converts the results into a reverse mapping from license names to package names I am wondering how useful this is as part of the generic API of `rospkg`. Where do you plan to use the new API? Depending on the answer why can't the logic be implemented there instead?

I couldn't come up with a better alternative direction, so I just share my usecase below.

I found this function convenient as a one-off to get all the list of licenses and all relevant packages (catkin and non-catkin packages), so that no other command is needed to get a complete list. Indeed, for those packages where license declaration wasn't found, I need additional steps (which is not part of this PR), but getting all the package names whose licenses couldn't be determined is a useful information IMO.

That said, with the current implementation in this PR, there's no room for capturing the license from the non-catkin packages (e.g. what if the upstream implementation changes so that license of non-catkin package can be obtained?). I'll think about updating this PR so that the code doesn't rule out the chance of getting license for non-catkin pkgs.

Reply:
I found my post doesn't make much sense so re-wrote as following:

----

From https://github.com/ros-infrastructure/rospkg/pull/133#discussion_r270968864
> Why should this function combine both - the results from `get_depends` as well as `get_rosdeps` - into a single result? That seems very use case specific.

The method `get_licenses` in this PR returns the dictionary of licenses and the packages. I found that useful that the output is not just a list of licenses but also the name of all packages in the dependency tree.

I found:
- `get_depends` returns the list of upstream catkin packages.
- `get_rosdeps` returns the list of upstream non-catkin packages.

In my usecase there's no reason to limit the scope of license introspection to catkin package (our customers want to know all the software installed on the delivered system). So including catking and non-catking packages actually important to me.

> In general since the function only converts the results into a reverse mapping from license names to package names I am wondering how useful this is as part of the generic API of `rospkg`. Where do you plan to use the new API? Depending on the answer why can't the logic be implemented there instead?

I found this function convenient as a one-off to get all the list of licenses and all relevant packages (catkin and non-catkin packages), so that no other command is needed to get a complete list. Since in `rospkg` there's already `get_depends`, which returns a set of pkgs in the dependency tree, what `get_licenses` adds is essentially a license layer to the `get_depends`'s output.

With the current implementation of this PR, capturing the licenses from the non-catkin packages is not included (e.g. what if the upstream implementation changes so that license of non-catkin package can be obtained?). I have a working code to capture licenses from system package (on Ubuntu) but due to the conversation with Dirk on previous PRs, I'm going with more granular PR.